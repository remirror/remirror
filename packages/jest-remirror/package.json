{
  "name": "jest-remirror",
  "version": "3.0.0-beta.6",
  "description": "A text editor for react built with prosemirror",
  "keywords": [
    "prosemirror",
    "react",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-remirror.js",
      "require": "./dist/jest-remirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-remirror-environment.js",
      "require": "./environment/dist/jest-remirror-environment.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/jest-remirror.cjs",
  "module": "./dist/jest-remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.6",
    "@remirror/pm": "3.0.0-beta.4",
    "@remirror/preset-core": "3.0.0-beta.6",
    "@testing-library/dom": "^9.3.0",
    "@types/sanitize-html": "^2.9.0",
    "jest-prosemirror": "3.0.0-beta.4",
    "sanitize-html": "^2.10.0",
    "tiny-warning": "^1.0.3"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "@types/node": "^16.18.34",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "@types/node": "*",
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    },
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-remirror.d.ts",
        "import": "./dist/jest-remirror.js",
        "require": "./dist/jest-remirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-remirror-environment.d.ts",
        "import": "./environment/dist/jest-remirror-environment.js",
        "require": "./environment/dist/jest-remirror-environment.cjs"
      },
      "./package.json": "./package.json"
    },
    "main": "./dist/jest-remirror.cjs",
    "module": "./dist/jest-remirror.js",
    "type": "module",
    "types": "./dist/jest-remirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts"
    ]
  }
}
