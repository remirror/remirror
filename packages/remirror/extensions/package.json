{
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/remirror-extensions.d.ts",
      "import": "./dist/remirror-extensions.js",
      "require": "./dist/remirror-extensions.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/remirror-extensions.cjs",
  "module": "./dist/remirror-extensions.js",
  "types": "./dist/remirror-extensions.d.ts"
}
