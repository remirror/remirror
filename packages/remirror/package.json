{
  "name": "remirror",
  "version": "3.0.0-beta.8",
  "description": "One editing package to rule them all, one editing package to bind them.",
  "keywords": [
    "prosemirror",
    "react",
    "remirror",
    "wysiwyg",
    "editing",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>",
    "ocavue "
  ],
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/remirror.js",
      "require": "./dist/remirror.cjs"
    },
    "./dom": {
      "types": "./dist-types/dom.d.ts",
      "import": "./dom/dist/remirror-dom.js",
      "require": "./dom/dist/remirror-dom.cjs"
    },
    "./extensions": {
      "types": "./dist-types/extensions.d.ts",
      "import": "./extensions/dist/remirror-extensions.js",
      "require": "./extensions/dist/remirror-extensions.cjs"
    },
    "./icons": {
      "types": "./dist-types/icons.d.ts",
      "import": "./icons/dist/remirror-icons.js",
      "require": "./icons/dist/remirror-icons.cjs"
    },
    "./package.json": "./package.json",
    "./styles/*": "./styles/*"
  },
  "main": "./dist/remirror.cjs",
  "module": "./dist/remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "dom",
    "extensions",
    "icons",
    "styles"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.8",
    "@remirror/core-constants": "3.0.0-beta.4",
    "@remirror/core-helpers": "4.0.0-beta.5",
    "@remirror/core-types": "3.0.0-beta.5",
    "@remirror/core-utils": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.8",
    "@remirror/extension-annotation": "3.0.0-beta.8",
    "@remirror/extension-bidi": "3.0.0-beta.8",
    "@remirror/extension-blockquote": "3.0.0-beta.8",
    "@remirror/extension-bold": "3.0.0-beta.8",
    "@remirror/extension-callout": "3.0.0-beta.8",
    "@remirror/extension-code": "3.0.0-beta.8",
    "@remirror/extension-code-block": "3.0.0-beta.8",
    "@remirror/extension-collaboration": "3.0.0-beta.8",
    "@remirror/extension-columns": "3.0.0-beta.8",
    "@remirror/extension-diff": "3.0.0-beta.8",
    "@remirror/extension-doc": "3.0.0-beta.8",
    "@remirror/extension-drop-cursor": "3.0.0-beta.8",
    "@remirror/extension-embed": "3.0.0-beta.8",
    "@remirror/extension-emoji": "3.0.0-beta.8",
    "@remirror/extension-entity-reference": "3.0.0-beta.8",
    "@remirror/extension-epic-mode": "3.0.0-beta.8",
    "@remirror/extension-events": "3.0.0-beta.8",
    "@remirror/extension-find": "1.0.0-beta.8",
    "@remirror/extension-font-family": "3.0.0-beta.8",
    "@remirror/extension-font-size": "3.0.0-beta.8",
    "@remirror/extension-gap-cursor": "3.0.0-beta.8",
    "@remirror/extension-hard-break": "3.0.0-beta.8",
    "@remirror/extension-heading": "3.0.0-beta.8",
    "@remirror/extension-history": "3.0.0-beta.8",
    "@remirror/extension-horizontal-rule": "3.0.0-beta.8",
    "@remirror/extension-image": "3.0.0-beta.8",
    "@remirror/extension-italic": "3.0.0-beta.8",
    "@remirror/extension-link": "3.0.0-beta.8",
    "@remirror/extension-list": "3.0.0-beta.8",
    "@remirror/extension-markdown": "3.0.0-beta.8",
    "@remirror/extension-mention": "3.0.0-beta.8",
    "@remirror/extension-mention-atom": "3.0.0-beta.8",
    "@remirror/extension-node-formatting": "3.0.0-beta.8",
    "@remirror/extension-paragraph": "3.0.0-beta.8",
    "@remirror/extension-placeholder": "3.0.0-beta.8",
    "@remirror/extension-positioner": "3.0.0-beta.8",
    "@remirror/extension-shortcuts": "3.0.0-beta.8",
    "@remirror/extension-strike": "3.0.0-beta.8",
    "@remirror/extension-sub": "3.0.0-beta.8",
    "@remirror/extension-sup": "3.0.0-beta.8",
    "@remirror/extension-tables": "3.0.0-beta.8",
    "@remirror/extension-text": "3.0.0-beta.8",
    "@remirror/extension-text-case": "3.0.0-beta.8",
    "@remirror/extension-text-color": "3.0.0-beta.8",
    "@remirror/extension-text-highlight": "3.0.0-beta.8",
    "@remirror/extension-trailing-node": "3.0.0-beta.8",
    "@remirror/extension-underline": "3.0.0-beta.8",
    "@remirror/extension-whitespace": "3.0.0-beta.8",
    "@remirror/icons": "3.0.0-beta.5",
    "@remirror/preset-core": "3.0.0-beta.8",
    "@remirror/preset-formatting": "3.0.0-beta.8",
    "@remirror/preset-wysiwyg": "3.0.0-beta.8",
    "@remirror/theme": "3.0.0-beta.5",
    "@types/refractor": "^3.0.2",
    "refractor": "^3.6.0"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.1",
    "@remirror/pm": "3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependencies": {
    "@remirror/pm": "^3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependenciesMeta": {
    "@types/prettier": {
      "optional": true
    },
    "prettier": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/remirror.d.ts",
        "import": "./dist/remirror.js",
        "require": "./dist/remirror.cjs"
      },
      "./dom": {
        "types": "./dom/dist/remirror-dom.d.ts",
        "import": "./dom/dist/remirror-dom.js",
        "require": "./dom/dist/remirror-dom.cjs"
      },
      "./extensions": {
        "types": "./extensions/dist/remirror-extensions.d.ts",
        "import": "./extensions/dist/remirror-extensions.js",
        "require": "./extensions/dist/remirror-extensions.cjs"
      },
      "./icons": {
        "types": "./icons/dist/remirror-icons.d.ts",
        "import": "./icons/dist/remirror-icons.js",
        "require": "./icons/dist/remirror-icons.cjs"
      },
      "./package.json": "./package.json",
      "./styles/*": "./styles/*"
    },
    "main": "./dist/remirror.cjs",
    "module": "./dist/remirror.js",
    "type": "module",
    "types": "./dist/remirror.d.ts"
  },
  "@remirror": {},
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "dom.ts",
      "extensions.ts",
      "icons.ts"
    ]
  }
}
