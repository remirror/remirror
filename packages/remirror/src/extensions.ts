// Extensions
export * from '@remirror/extension-annotation';
export * from '@remirror/extension-bidi';
export * from '@remirror/extension-blockquote';
export * from '@remirror/extension-bold';
export * from '@remirror/extension-callout';
export * from '@remirror/extension-code';
export * from '@remirror/extension-code-block';
export * from '@remirror/extension-codemirror5';
export * from '@remirror/extension-collaboration';
export * from '@remirror/extension-columns';
export * from '@remirror/extension-diff';
export * from '@remirror/extension-doc';
export * from '@remirror/extension-drop-cursor';
export * from '@remirror/extension-embed';
export * from '@remirror/extension-emoji';
export * from '@remirror/extension-entity-reference';
export * from '@remirror/extension-epic-mode';
export * from '@remirror/extension-events';
export * from '@remirror/extension-find';
export * from '@remirror/extension-font-family';
export * from '@remirror/extension-font-size';
export * from '@remirror/extension-gap-cursor';
export * from '@remirror/extension-hard-break';
export * from '@remirror/extension-heading';
export * from '@remirror/extension-history';
export * from '@remirror/extension-horizontal-rule';
export * from '@remirror/extension-image';
export * from '@remirror/extension-italic';
export * from '@remirror/extension-link';
export * from '@remirror/extension-list';
export * from '@remirror/extension-markdown';
export * from '@remirror/extension-mention';
export * from '@remirror/extension-mention-atom';
export * from '@remirror/extension-node-formatting';
export * from '@remirror/extension-paragraph';
export * from '@remirror/extension-placeholder';
export * from '@remirror/extension-positioner';
export * from '@remirror/extension-shortcuts';
export * from '@remirror/extension-strike';
export * from '@remirror/extension-sub';
export * from '@remirror/extension-sup';
export * from '@remirror/extension-tables';
export * from '@remirror/extension-text';
export * from '@remirror/extension-text-case';
export * from '@remirror/extension-text-color';
export * from '@remirror/extension-text-highlight';
export * from '@remirror/extension-trailing-node';
export * from '@remirror/extension-underline';
export * from '@remirror/extension-whitespace';
export * from '@remirror/extension-yjs';

// Presets
export * from '@remirror/preset-core';
export * from '@remirror/preset-formatting';
export * from '@remirror/preset-wysiwyg';
