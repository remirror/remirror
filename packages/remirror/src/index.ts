export * from '@remirror/core';
export * from '@remirror/theme';
