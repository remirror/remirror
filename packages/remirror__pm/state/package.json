{
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/remirror-pm-state.d.ts",
      "import": "./dist/remirror-pm-state.js",
      "require": "./dist/remirror-pm-state.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/remirror-pm-state.cjs",
  "module": "./dist/remirror-pm-state.js",
  "types": "./dist/remirror-pm-state.d.ts"
}
