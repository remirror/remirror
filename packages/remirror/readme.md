# remirror

> One package to rule them all, one entry point to bind them.

[![Version][version]][npm] [![Weekly Downloads][downloads-badge]][npm] [![Bundled size][size-badge]][size] [![Typed Codebase][typescript]](#) [![MIT License][license]](#)

[version]: https://flat.badgen.net/npm/v/remirror
[npm]: https://npmjs.com/package/remirror
[license]: https://flat.badgen.net/badge/license/MIT/purple
[size]: https://bundlephobia.com/result?p=remirror
[size-badge]: https://flat.badgen.net/bundlephobia/minzip/remirror
[typescript]: https://flat.badgen.net/badge/icon/TypeScript?icon=typescript&label
[downloads-badge]: https://badgen.net/npm/dw/remirror/red?icon=npm

## Installation

```bash
# yarn
yarn add remirror @remirror/pm

# pnpm
pnpm add remirror @remirror/pm

# npm
npm install remirror @remirror/pm
```

The `remirror` package will automatically install the `@remirror/pm` package for you. You only need to install it yourself if you'd like to ensure consistent versions of the `prosemirror-*` libraries when importing from `@remirror/pm/state` instead of `prosemirror-state` or `@remirror/pm/model` instead of `prosemirror-model`.

## Usage

Rather than installing multiple scoped packages, the `remirror` package is a gateway to using all the goodness that remirror provides while minimising your bundle size.

The following creates a controlled editor with React.

```tsx
import React from 'react';
import { SocialPreset } from 'remirror/extensions';
import { Remirror, SocialEmojiComponent, useRemirror } from '@remirror/react';

const EditorWrapper = () => {
  const socialPreset = new SocialPreset();
  const { state, onChange } = useRemirror({ extensions: () => [...socialPreset()] });

  return (
    <Remirror state={state} onChange={onChange} manager={manager} autoRender={true}>
      <SocialEmojiComponent />
    </Remirror>
  );
};
```

These are the entry points available through the `remirror` package.

- `remirror` - All the core functionality available through `@remirror/core`.
- `remirror/extensions` - All the core extensions and presets made available through the main `remirror` repository. This doesn't include any framework specific extensions and presets.
- `remirror/dom` - The dom framework implementation of via `createDomEditor`.
