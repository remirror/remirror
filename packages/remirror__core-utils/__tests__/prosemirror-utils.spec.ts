import { jest } from '@jest/globals';
import {
  atomBlock,
  atomContainer,
  atomInline,
  blockquote,
  createEditor,
  doc,
  h2,
  p,
  schema,
  table,
  td,
  tdCursor,
  tdEmpty,
  tr as row,
} from 'jest-prosemirror';
import { renderEditor } from 'jest-remirror';
import { BoldExtension, createCoreManager, HeadingExtension } from 'remirror/extensions';
import { omit } from '@remirror/core-helpers';
import { NodeSelection, Selection, TextSelection } from '@remirror/pm/state';

import {
  cloneTransaction,
  composeTransactionSteps,
  containsAttributes,
  findElementAtPosition,
  findNodeAtSelection,
  findParentNode,
  findParentNodeOfType,
  findPositionOfNodeAfter,
  findSelectedNodeOfType,
  hasTransactionChanged,
  isNodeActive,
  isNodeOfType,
  isSelectionEmpty,
  removeNodeAtPosition,
  schemaToJSON,
} from '../';

describe('isNodeOfType', () => {
  it('matches with a singular nodeType', () => {
    expect(isNodeOfType({ types: schema.nodes.paragraph, node: p() })).toBeTrue();
  });

  it('matches with an array of nodeTypes', () => {
    const { paragraph, blockquote: bq } = schema.nodes;

    expect(isNodeOfType({ types: [paragraph, bq], node: blockquote() })).toBeTrue();
  });
});

describe('removeNodeAtPosition', () => {
  it('removes block top level nodes at specified position', () => {
    const {
      state: { tr },
    } = createEditor(doc(p('x'), p('one')));
    removeNodeAtPosition({ pos: 3, tr });

    expect(tr.doc).toEqualProsemirrorNode(doc(p('x')));
  });

  it('removes nested inline nodes at specified position', () => {
    const {
      state: { tr },
    } = createEditor(doc(p('one', atomInline())));
    removeNodeAtPosition({ pos: 4, tr });

    expect(tr.doc).toEqualProsemirrorNode(doc(p('one')));
  });
});

describe('findPositionOfNodeAfter', () => {
  it('returns `undefined` when none exists', () => {
    const {
      state: { selection },
    } = createEditor(doc(p('<cursor>')));
    const result = findPositionOfNodeAfter(selection);

    expect(result).toBeUndefined();
  });

  it('supports tables', () => {
    const node = table(row(tdEmpty), row(tdEmpty));
    const { state } = createEditor(doc(p('abcd'), node));
    const position = findPositionOfNodeAfter(state.doc.resolve(6));

    expect(position).toEqual({ pos: 6, start: 7, end: 20, depth: 1, node });
  });

  it('supports blockquotes', () => {
    const node = blockquote(p(''));
    const { state } = createEditor(doc(p('abcd'), node));
    const position = findPositionOfNodeAfter(state.doc.resolve(6));

    expect(position).toEqual({ pos: 6, start: 7, end: 10, depth: 1, node });
  });

  it('supports nested leaf nodes', () => {
    const node = atomBlock();
    const {
      state: { selection },
    } = createEditor(doc(p('abcd'), table(row(td(p('1'), '<cursor>', node)))));
    const position = findPositionOfNodeAfter(selection);

    expect(position).toEqual({ pos: 12, start: 13, end: 13, depth: 4, node });
  });

  it('supports non-nested leaf nodes', () => {
    const node = atomBlock();
    const { state } = createEditor(doc(p('abcd'), node));
    const position = findPositionOfNodeAfter(state.doc.resolve(6));

    expect(position).toEqual({ pos: 6, start: 7, end: 7, depth: 1, node });
  });

  it('supports leaf nodes with with nested inline atom nodes', () => {
    const node = atomContainer(atomBlock());
    const { state } = createEditor(doc(p('abcd'), node));
    const position = findPositionOfNodeAfter(state.doc.resolve(6));

    expect(position).toEqual({ pos: 6, start: 7, end: 9, depth: 1, node });
  });
});

describe('findElementAtPosition', () => {
  it('should return DOM reference of a top level block leaf node', () => {
    const { view } = createEditor(doc(p('text'), atomBlock()));
    const reference = findElementAtPosition(6, view);

    expect(reference instanceof HTMLDivElement).toBe(true);
    expect(reference.getAttribute('data-node-type')).toBe('atomBlock');
  });

  it('should return DOM reference of a nested inline leaf node', () => {
    const { view } = createEditor(doc(p('one', atomInline(), 'two')));
    const reference = findElementAtPosition(4, view);

    expect(reference instanceof HTMLSpanElement).toBe(true);
    expect(reference.getAttribute('data-node-type')).toBe('atomInline');
  });

  it('should return DOM reference of a content block node', () => {
    const { view } = createEditor(doc(p('one'), blockquote(p('two'))));
    const reference = findElementAtPosition(5, view);

    expect(reference instanceof HTMLQuoteElement).toBe(true);
  });

  it('should return DOM reference of a text node when offset=0', () => {
    const { view } = createEditor(doc(p('text')));
    const reference = findElementAtPosition(1, view);

    expect(reference instanceof HTMLParagraphElement).toBe(true);
  });

  it('should return DOM reference of a paragraph if cursor is inside of a text node', () => {
    const { view } = createEditor(doc(p(atomInline(), 'text')));
    const reference = findElementAtPosition(3, view);

    expect(reference instanceof HTMLParagraphElement).toBe(true);
  });
});

describe('selectionEmpty', () => {
  it('returns false when selection is active', () => {
    const {
      state: { selection },
    } = createEditor(doc(p('inline'), p('<start>aba<end>', 'awesome')));

    expect(isSelectionEmpty(selection)).toBeFalse();
  });

  it('returns true when cursor is active', () => {
    const {
      state: { selection },
    } = createEditor(doc(p('inline'), p('<cursor>aba')));

    expect(isSelectionEmpty(selection)).toBeTrue();
  });

  it('returns true when no cursor', () => {
    const {
      state: { selection },
    } = createEditor(doc(p('inline'), p('aba')));

    expect(isSelectionEmpty(selection)).toBeTrue();
  });
});

describe('transactionChanged', () => {
  it('returns false when neither text nor state has change', () => {
    const {
      state: { tr },
      view,
    } = createEditor(doc(p('inline'), p('<start>aba<end>', 'awesome')));
    view.dispatch(tr);

    expect(hasTransactionChanged(tr)).toBeFalse();
  });

  it('returns true when the doc has changed', () => {
    const {
      state: { tr },
    } = createEditor(doc(p('inline'), p('<start>aba<end>', 'awesome')));
    tr.deleteSelection();

    expect(hasTransactionChanged(tr)).toBeTrue();
  });

  it('returns true when cursor changes', () => {
    const { state } = createEditor(doc(p('inline'), p('aba<cursor>')));
    const tr = state.tr.setSelection(TextSelection.atStart(state.doc));

    expect(hasTransactionChanged(tr)).toBeTrue();
  });
});

describe('cloneTransaction', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('clones the transaction', async () => {
    await new Promise<void>((resolve) => {
      const {
        state: { tr },
      } = createEditor(doc(p()));

      setTimeout(() => {
        const clonedTr = cloneTransaction(tr);

        expect(omit(tr, ['time'])).toEqual(omit(clonedTr, ['time']));

        resolve();
      }, 10);

      jest.advanceTimersByTime(20);
    });
  });
});

describe('composeTransactionSteps', () => {
  it('combines the steps of the given transactions', () => {
    const { state } = createEditor(doc(p('one'), p('two')));
    const tr1 = cloneTransaction(state.tr.replaceRangeWith(0, 5, p('ONE')));
    const tr2 = cloneTransaction(state.tr.replaceRangeWith(5, 10, p('TWO')));

    const composedTr = composeTransactionSteps([tr1, tr2], state);
    expect(composedTr.steps).toMatchInlineSnapshot(`
      [
        {
          "from": 0,
          "slice": {
            "content": [
              {
                "content": [
                  {
                    "text": "ONE",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
          },
          "stepType": "replace",
          "to": 5,
        },
        {
          "from": 5,
          "slice": {
            "content": [
              {
                "content": [
                  {
                    "text": "TWO",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
          },
          "stepType": "replace",
          "to": 10,
        },
      ]
    `);
    expect(composedTr.doc).toEqualProsemirrorNode(doc(p('ONE'), p('TWO')));
  });
});

describe('findParentNode', () => {
  it('finds parent node with cursor directly inside', () => {
    const node = p('hello <cursor>');
    const {
      state: { selection },
    } = createEditor(doc(node));
    const position = findParentNode({
      predicate: (pmNode) => pmNode.type === schema.nodes.paragraph,
      selection,
    });

    expect(position).toEqual({ pos: 0, start: 1, end: 8, depth: 1, node });
  });

  it('finds parent from inside nested child', () => {
    const {
      state: { selection },
    } = createEditor(doc(table(row(tdCursor))));
    const result = findParentNode({
      predicate: (pmNode) => pmNode.type === schema.nodes.table,
      selection,
    });

    expect(result?.node.type.name).toBe('table');
  });

  it('returns `undefined` when no parent node found', () => {
    const {
      state: { selection },
    } = createEditor(doc(table(row(tdCursor))));
    const result = findParentNode({
      predicate: (pmNode) => pmNode.type === schema.nodes.table_header,
      selection,
    });

    expect(result).toBeUndefined();
  });
});

describe('findParentNodeOfType', () => {
  it('finds parent node of a given `nodeType`', () => {
    const node = p('hello <cursor>');
    const {
      state: { selection },
    } = createEditor(doc(node));
    const position = findParentNodeOfType({ types: schema.nodes.paragraph, selection });

    expect(position).toEqual({ pos: 0, start: 1, end: 8, depth: 1, node });
  });

  it('returns `undefined` when given `nodeType` is not found', () => {
    const {
      state: { selection },
    } = createEditor(doc(p('hello <cursor>')));
    const result = findParentNodeOfType({ types: schema.nodes.table, selection });

    expect(result).toBeUndefined();
  });

  it('finds parent node when `nodeType` is an array', () => {
    const node = p('hello <cursor>');
    const {
      state: {
        schema: {
          nodes: { paragraph, blockquote: bq, table: table_ },
        },
        selection,
      },
    } = createEditor(doc(node));
    const position = findParentNodeOfType({ types: [table_, bq, paragraph], selection });

    expect(position).toEqual({ pos: 0, start: 1, end: 8, depth: 1, node });
  });
});

describe('nodeActive', () => {
  it('shows active when within an active region', () => {
    const { state, schema: sch } = createEditor(
      doc(p('Something', blockquote(p('is <cursor>in blockquote')))),
    );

    expect(isNodeActive({ state, type: sch.nodes.blockquote })).toBeTrue();
  });

  it('returns false when not within the node', () => {
    const { state, schema: sch } = createEditor(doc(p('Something<cursor>', blockquote('hello'))));

    expect(isNodeActive({ state, type: sch.nodes.blockquote })).toBeFalse();
  });

  it('returns false with text selection surrounds the node', () => {
    const { state, schema: sch } = createEditor(
      doc(p('Something<start>', blockquote('is italic'), '<end> here')),
    );

    expect(isNodeActive({ state, type: sch.nodes.blockquote })).toBeFalse();
  });

  it('returns true when node selection directly before node', () => {
    const { state, schema: sch } = createEditor(
      doc(p('Something', blockquote('<node>is italic'), 'here')),
    );

    expect(isNodeActive({ state, type: sch.nodes.blockquote })).toBeTrue();
  });

  it('returns false nested within other nodes', () => {
    const { state, schema: sch } = createEditor(
      doc(p('a<node>', p(p(blockquote('is italic')), 'here'))),
    );

    expect(isNodeActive({ state, type: sch.nodes.blockquote })).toBeFalse();
  });

  it('matches nodes by specified attributes', () => {
    const { state, schema: sch } = createEditor(
      doc(p('Something'), h2('level <cursor> heading'), p('here')),
    );

    expect(isNodeActive({ state, type: sch.nodes.heading, attrs: { level: 1 } })).toBeFalse();
    expect(isNodeActive({ state, type: sch.nodes.heading, attrs: { level: 2 } })).toBeTrue();
  });

  it('matches partial attributes', () => {
    const { nodes, add, attributeNodes } = renderEditor([
      new HeadingExtension({ extraAttributes: { custom: { default: 'custom' } } }),
    ]);

    const { doc, p } = nodes;
    const hCustom = attributeNodes.heading({ custom: 'test', level: 2 });
    const { state, schema } = add(
      doc(p('Something'), hCustom('level <cursor> heading'), p('here')),
    );
    const type = schema.nodes.heading;

    expect(isNodeActive({ state, type, attrs: { level: 1 } })).toBeFalse();
    expect(isNodeActive({ state, type, attrs: { level: 2 } })).toBeTrue();
    expect(isNodeActive({ state, type, attrs: { level: 2, custom: 'test' } })).toBeTrue();
    expect(isNodeActive({ state, type, attrs: { level: 2, custom: 'no' } })).toBeFalse();
  });
});

describe('findSelectedNodeOfType', () => {
  it('should return `undefined` if selection is not a NodeSelection', () => {
    const {
      state: { selection },
    } = createEditor(doc(p('<cursor>')));
    const node = findSelectedNodeOfType({ types: schema.nodes.paragraph, selection });

    expect(node).toBeUndefined();
  });

  it('should return selected node of a given `nodeType`', () => {
    const { state } = createEditor(doc(p('<cursor>one')));
    const tr = state.tr.setSelection(NodeSelection.create(state.doc, 0));
    const selectedNode = findSelectedNodeOfType({
      types: state.schema.nodes.paragraph,
      selection: tr.selection,
    });

    expect(selectedNode!.node.type.name).toBe('paragraph');
  });

  it('should return selected node of one of the given `nodeType`s', () => {
    const { state } = createEditor(doc(p('<node>one')));
    const { paragraph, table: t } = state.schema.nodes;
    const selectedNode = findSelectedNodeOfType({
      types: [paragraph, t],
      selection: state.selection,
    });

    expect(selectedNode?.node.type.name).toBe('paragraph');
  });
});

describe('findNodeAt...', () => {
  const expectedEnd = h2('Heading here');
  const expectedStart = p('<cursor> You am champion');
  const pmDocument = doc(expectedStart, expectedEnd);

  it('findNodeAtSelection', () => {
    const selection = Selection.atEnd(pmDocument);
    const { node, pos, start } = findNodeAtSelection(selection);

    expect(node).toBe(expectedEnd);
    expect(pos).toBe(18);
    expect(start).toBe(19);
  });
});

test('schemaToJSON', () => {
  expect(schemaToJSON(createCoreManager([new BoldExtension()]).schema)).toMatchSnapshot();
});

describe('containsAttributes', () => {
  it('return true when node contains supplied attribute values', () => {
    expect(containsAttributes(h2(), { level: 2 })).toBeTrue();
  });

  it('return false when node does not contains supplied attribute values', () => {
    expect(containsAttributes(h2(), { level: 1 })).toBeFalse();
    expect(containsAttributes(h2(), { level: 2, something: 'else' })).toBeFalse();
  });

  it('return false when node does not have any attribute values', () => {
    expect(containsAttributes(p(), { level: 1 })).toBeFalse();
  });

  it('return true when mark contains supplied attribute values', () => {
    const link = schema.marks.link.create({ href: '//test.com', title: 'foo' });
    expect(containsAttributes(link, { href: '//test.com' })).toBeTrue();
    expect(containsAttributes(link, { href: '//test.com', title: 'foo' })).toBeTrue();
  });

  it('return false when mark does not contains supplied attribute values', () => {
    const link = schema.marks.link.create({ href: '//remirror.io', title: 'foo' });
    expect(containsAttributes(link, { href: '//prosemirror.net' })).toBeFalse();
    expect(containsAttributes(link, { href: '//remirror.io', title: 'bar' })).toBeFalse();
    expect(
      containsAttributes(link, { href: '//remirror.io', title: 'foo', something: 'else' }),
    ).toBeFalse();
  });

  it('return false when mark does not have any attribute values', () => {
    const strong = schema.marks.strong.create();
    expect(containsAttributes(strong, { href: '//remirror.io' })).toBeFalse();
  });
});
