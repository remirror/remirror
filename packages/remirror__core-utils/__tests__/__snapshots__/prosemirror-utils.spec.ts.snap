// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaToJSON 1`] = `
{
  "marks": {
    "bold": {
      "attrs": {},
      "group": "formattingMark fontStyle",
      "parseDOM": [
        {
          "getAttrs": [Function],
          "tag": "strong",
        },
        {
          "getAttrs": [Function],
          "tag": "b",
        },
        {
          "getAttrs": [Function],
          "style": "font-weight",
        },
      ],
      "toDOM": [Function],
    },
  },
  "nodes": {
    "doc": {
      "attrs": {},
      "content": "block+",
      "group": undefined,
    },
    "paragraph": {
      "attrs": {},
      "content": "inline*",
      "draggable": false,
      "group": "lastNodeCompatible textBlock block formattingNode",
      "parseDOM": [
        {
          "getAttrs": [Function],
          "tag": "p",
        },
      ],
      "toDOM": [Function],
    },
    "text": {
      "group": "inline",
    },
  },
}
`;
